nautilus_core/common/src/throttler/mod.rs

```diff
@@ -0,0 +1,457 @@
+// -------------------------------------------------------------------------------------------------
+//  Copyright (C) 2015-2024 Nautech Systems Pty Ltd. All rights reserved.
+//  https://nautechsystems.io
+//
+//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
+//  You may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+// -------------------------------------------------------------------------------------------------
+
+use std::{cmp::max, collections::VecDeque};
+
+use nautilus_core::nanos::UnixNanos;
+
+use crate::{clock::Clock, timer::TimeEvent};
+
+type ThrottlerCallbackFn<T> = dyn Fn(&mut Throttler<T>, TimeEvent);
+
+// New struct to hold the callback
+pub struct ThrottlerCallback<T> {
+    inner: Box<ThrottlerCallbackFn<T>>,
+}
+
+impl<T> ThrottlerCallback<T> {
+    pub fn new<F>(f: F) -> Self
+    where
+        F: Fn(&mut Throttler<T>, TimeEvent) + 'static,
+    {
+        ThrottlerCallback { inner: Box::new(f) }
+    }
+
+    pub fn call(&self, throttler: &mut Throttler<T>, event: TimeEvent) {
+        (self.inner)(throttler, event);
+    }
+}
+
+/// A throttler that limits the rate at which messages are sent.
+pub struct Throttler<T> {
+    /// The number of messages received.
+    pub recv_count: usize,
+    /// The number of messages sent.
+    pub sent_count: usize,
+    /// Whether the throttler is currently limiting the message rate.
+    pub is_limiting: bool,
+    /// The maximum number of messages that can be sent within the interval.
+    pub limit: usize,
+    /// The buffer of messages to be sent.
+    buffer: VecDeque<T>,
+    /// The timestamps of the sent messages.
+    timestamps: VecDeque<UnixNanos>,
+    /// Whether the throttler has warmed up.
+    warm: bool,
+    /// The interval between messages in nanoseconds.
+    interval: u64,
+    /// The clock used to keep track of time.
+    clock: Box<dyn Clock>,
+    /// The name of the timer.
+    timer_name: String,
+    /// The function to send a message.
+    output_send: fn(T),
```

Comment:
`fn` style function pointers are good for testing but isn't flexible enough to handle other use cases because it can't store data[^1]. Typical use case would be sending the message across a message bus for which the function will need to store a reference to the message bus. This can perhaps be done using a closure too, by "move"ing the message bus reference bus inside it. Perhaps, using a Fn style trait here will be more flexible. Something like the ones defined for MessageBusHandler and the likes.

[^1]: https://stackoverflow.com/a/64298764